### ISSUE TITLE
PHPMD complaining about UnusedFormalParameters when implementing an interface.

### ISSUE THREAD

**Issue by heiglandreas** [phpmd/phpmd#515]
Currently PHPMD (as far as I can see it) complains about unused formal parameters when implementing an interface.

Given the interface declares a method ```public function foo ($bar, $baz);``` and my implementation only needs ```$bar``` to execute I will always see the ```Avoid unused parameters such as '$baz'.```-message. 

Adding the ```@inheritdoc```-workaround is a… well… workaround and doesn'T solve the underlying issues that the check shouldn't be applied to methods implementing an abstract method.

**Comment by ravage84** [comment]
@heiglandreas so your suggestion is to never check for unused parameters when the class is implementing an abstract method?

This would certainly generate lots of blind spots. Selectively [suppressing the warning](https://phpmd.org/documentation/suppress-warnings.html) would be a better fit, I guess.

**Comment by heiglandreas** [comment]
I‘d only skip the check when a function implements an abstract method. That should never create a blind spot as the definition of the method signature has been done elsewhere and the concrete implementation is merely an implememtation detail. 

Suppressing the warning might be a nice workaround, but as the issue is that the definition of the method signature was done somewhere else where I don‘t have an influence on, the warning IMHO is a false positive that should be avoided in the first place.

**Comment by ravage84** [comment]
> That should never create a blind spot as the definition of the method signature has been done elsewhere and the concrete implementation is merely an implememtation detail.

The unused parameter rule does not inspect the definition but the usage of a method and its signature. 
If you have a method but do not use one of its parameter then that can be a hint that you have a bad or at least a too general design. Often this happens with classes/methods from frameworks.
Thus ignoring unused parameters in methods that implement an abstract method will still lead to blind spots.

So, deciding case by case which violations for which methods is a more conscious method. Cause this way you are fully aware of your less shiny places in your code.

**Comment by kylekatarnls** [comment]
By marking an argument as mandatory, the interface expect you to use it. If the interface is properly done, and does allow to omit `$bar` and `$baz`, it means they both should be taken into account to implement `foo()`.

Either, the interface should relax: `public function foo($bar = null, $baz = null);` so your explicitly not required to use them. Or it's the sign your implementation of the method is actually incomplete.

So I would also opt for a manual muting for each violation, but not for an automatic ignore of unused parameter required by interfaces.

**Comment by heiglandreas** [comment]
The interface might expect me to use it. But it is my implementation that decides whether to use it due to a special use case. So there is no way that "my implementation is incomplete" as that is entirely an implementation detail. 

To illustrate that a small example.

Imagine a PSR15 based middleware that shall shortcut the Middleware-Pipeline in case of maintenance. So I would implement a MaintenanceMiddleware that will be dynamically injected to the first place of the middleware-piepline under certain circumstances.

So I could write the following code:

```php
class MaintenaceMiddleware
{
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        return new MaintenanceResponse('Sorry, we'Re down for maintenance);
    }
}
```

Perfectly valid implementation of the middleware. Under the given circumstances this middleware does not care about the Request or the RequestHandler. But that is an implementation detail the interface does not care about.

But in this case PHPMD would complain that the parameters weren't used.

And btw: Just because a parameter is nullable does not mean that it should not be used. It only means that the parameter does not need ti be passed to the function/method in question. It should – according to PHPMD's current logic – still be used because the parameter is defined .

**Comment by kylekatarnls** [comment]
Yep indeed. Relevant.

**Comment by ravage84** [comment]
@heiglandreas in your PSR-15 example, it would take one line to suppress the warning. 
I agree, sometimes, like in your example, you simply don't need all or any parameter. But in such cases add a suppression and good you are. This also documents the specific intent not to use the parameters.

**Comment by tvbeek** [comment]
I think we need to be clear that the rules are not holy rules. It is possible to ignore it on a specific position. In my team we do that and write why so it is clear for other people.

I think the example from @heiglandreas is a good example about a place to ignore it. Maybe we need to take a look at the documentation to see how we can make it clear that it is possible to ignore rules on specific positions.

**Comment by ravage84** [comment]
By the way, this issue was previously discussed in the following issues:

https://github.com/phpmd/phpmd/issues/24
https://github.com/phpmd/phpmd/issues/50
https://github.com/phpmd/phpmd/issues/180

Also, there is a fugly workarouond with the ``@inheritdoc`` annotation:

https://github.com/phpmd/phpmd/commit/d033354f43bfac5b06146c20cb2ef0aff9196ada#src/main/php/PHP/PMD/Rule/UnusedFormalParameter.php

**Comment by ravage84** [comment]
How about we add a ``ignored_interfaces`` (or similar) property?

With that, one could either ignore full classes/interfaces, methods of a single class/interface or methods of any class/interface by name

For example:

- ``Controller`` would ignore the rule for any method of this class/interface.
- ``Controller::beforeRender`` would ignore the rule for this method of that class/interface
- ``beforeRender`` would ignore the rule for this method of any class/interface

**Comment by hackel** [comment]
A similar concept is when defining a closure that is expected to conform to a particular interface (although not formally defined), there are often cases where an earlier argument is not required and should not be reported.  e.g.
```php
array_filter(['bad', 'good'], fn ($value, $key) => $key % 2, ARRAY_FILTER_USE_BOTH);
```
In this case, it should not generate the `Avoid unused parameters such as '$value'` error.

**Comment by kylekatarnls** [comment]
I see the point, but here using `ARRAY_FILTER_USE_KEY` would be way better.

While it's not always immediately and easily fixable, having a not used parameter should warn a flag. First: from inside the function you don't know what it cost to give you this parameter you don't use.

See this is example:

```php
interface UserChercker
{
    public function check(string $name, string $id): bool;
}

class SimpleAppChecker implements UserChercker
{
    public function check(string $name, string $id): bool
    {
        // In my simple app, there is no constraint on $name
        return strlen($id) >= 3;
    }
}

// Somewhere else we call it:
/** @var UserChercker */
$checker = getChecker(); // Maybe it's a SimpleAppChecker, may an other implementation

// If now I want to call check() I need to fulfill both arguments,
// assuming I only have the ID for now, I will calculate the $name
if ($checker->check(getNameFromId($id), $id)) {
    echo 'checked';
}
```

Here the big problem, is `getNameFromId()` might call a SQL query to retrieve a user while we actually just need the name.

So when having `$name` unused feels like you have no choice because the interface (same for callback/closures signatures) forces you to get it, it's actually the sign the interface does too much for your case. It means you need a more precise interface:

```php
interface SimplerUserChercker
{
    public function simpleCheck(string $id): bool;
}

$checker = getChecker(); // may implements UserChercker, SimplerUserChercker, or both

$checked = $checker instanceof SimplerUserChercker
    ? $checker->simpleCheck($id)
    : $checker->check(getNameFromId($id), $id)

if ($checked) {
    echo 'checked';
}
```

This way `$name` is never unused and calculated only if needed.

And the same way, there is `ARRAY_FILTER_USE_KEY` in array_filter, there is often an alternative to not call the functions that passes unused parameters. If it's missing in a library you're using, you might rise it as an optimization to the library maintainer.

**Comment by heiglandreas** [comment]
The issue becomes anoying when you implement an interface that you have no control over. 

For example when implementing a [PSR-15 Middleware](https://www.php-fig.org/psr/psr-15/#22-psrhttpservermiddlewareinterface) that returns a maintenance page I would write the [already described code](https://github.com/phpmd/phpmd/issues/515#issuecomment-544139156) like this:

```php
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;

class MaintenaceMiddleware implements MiddlewareInterface
{
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        return new MaintenanceResponse('Sorry, we\'re down for maintenance');
    }
}
```

In cases like this I'm not able to modify the interface but PHPMD will always yell at me that I have forgotten to use one of the parameters.

**Comment by kylekatarnls** [comment]
> you implement an interface that you have no control over

That's the point. As I said, if you use nothing of an interface, you should probably not use this interface in this situation.

You're in maintenance mode, so what's the point in waiting for `$request` and `$handler` to be created? You very likely can enter a maintenance from an earlier entry-point.

**Comment by ravage84** [comment]
@kylekatarnls  I would be OK to add a property to ignore certain classes or methods. 

> How about we add a `ignored_interfaces` (or similar) property?
> 
> With that, one could either ignore full classes/interfaces, methods of a single class/interface or methods of any class/interface by name
> 
> For example:
> 
>   * `Controller` would ignore the rule for any method of this class/interface.
> 
>   * `Controller::beforeRender` would ignore the rule for this method of that class/interface
> 
>   * `beforeRender` would ignore the rule for this method of any class/interface

**Comment by heiglandreas** [comment]
@kylekatarnls Sometimes you need to implement an interface to be able to use other "magic". Like to add a Middleware to a MiddlewareStack. As for example adding a Middleware to a MiddlewareStack *requires* you to implement the `MiddlewareInterface`. And as that is handled in external code not under your control you can either re-implement it yourself or decide to deliberately *not* use the parameters the interface offers you.

Besides that: An interface *offers*  parameters. The implementation decides whether to *use* them or not.

**Comment by ravage84** [comment]
I or one of my colleagues are going to open a PR for my proposed ignore/exception list property.

**Comment by koenhendriks** [comment]
It would be great if we could create some exceptions for certain variables used by for example, Laravel.

```xml
<rule ref="rulesets/unusedcode.xml/UnusedFormalParameter">
        <properties>
            <property name="exceptions" value="request" />
        </properties>
    </rule>
```

Currently this is possible for `UnusedLocalVariable` but not for `UnusedFormalParameter` (see https://phpmd.org/rules/unusedcode.html)

**Comment by AJenbo** [comment]
`$request` is optional, if you don't use it don't hit it in your controller methods. It's just there in the template as it is often used, but if you do not use it you should remove it.

**Comment by koenhendriks** [comment]
> `$request` is optional, if you don't use it don't hit it in your controller methods. It's just there in the template as it is often used, but if you do not use it you should remove it.

@AJenbo Yes thats correct for your example in the controller methods, but this will not work for a Laravel resource that extends the `JsonResource`. The `toArray` is overwritten to create the resource and [it expects](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Http/Resources/Json/JsonResource.php#L115) a `$request`.

So in this example, building an API with Laravel for, lets say 20 resources means I have to `@SuppressWarnings(UnusedFormalParameter)` on every resource that doesn't use the request and every future API resource I'll build.

And that's just Laravel ofcourse. PHPMD is a standalone tool for lots of projects with any kind of framework. It would be useful to setup a rule with custom parameter names for when you extend classes or implement interfaces from third party vendors.

**Comment by kylekatarnls** [comment]
Yes, `<property name="exceptions" value="request" />` could be an easy shortcut. But it's still hiding misery. The clean fix would be to ignore parameter from inherited methods.

Or even better mark only 1 parameter of 1 method of 1 interface as an exception for a finest granularity. It would make sense IMO because this is the actual code smell.

If we take your example, you have 20 resources receiving `$request` and none of them use it (and `$request` is clearly not a cheap object in Laravel), this also assume that `JsonResource::toArray` can only be used in some HTTP context, while you could technically use JSON in CLI, sockets, etc. So IMHO, [this line you mention](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Http/Resources/Json/JsonResource.php#L115) is actually a not-ideal pattern that a mess detector is legit to warn about. It should at least make optional `?Request $request = null`

**Comment by koenhendriks** [comment]
> So IMHO, [this line you mention](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Http/Resources/Json/JsonResource.php#L115) is actually a not-ideal pattern that a mess detector is legit to warn about. It should at least make optional ?Request $request = null

@kylekatarnls Yes I agree. But the thing is that this is inside a third party vendors code. So I would like to ignore it as I can't (realistically, without MR to the project) fix this 'not-ideal pattern'. I could create my own `JsonResource` class and create signatures that I like, but that kind of defeats the purpose of using a framework.

**Comment by koenhendriks** [comment]
@tvbeek Yes thats an easy workaround for a single file, but what if you have 20 resources? It will become tedious work to suppress each method in each file. 

I agree with @kylekatarnls that the clean fix would be to ignore parameter from inherited methods that are outside our own application namespace (so 3rd party vendor classes).

**Comment by darius-v** [comment]
supressing makes developers confused if they are doing right or wrong by supressing. You can always supress something because you think this is false message. But you might thing in wrong way. As someone said - maybe you do not need that interface. So I think when mess detector shows error, it really has to be an error, not "maybe". PhpStorm by the way also detects unsued parameters , but it does not show error when they the interface requires them.

**Comment by kylekatarnls** [comment]
> it really has to be an error, not "maybe"

No tool should be trusted blindly, whatever the static analyzer you use, they all have many issues open yet to be solved, each new PHP version / new feature / new coding pattern may come with incompatibilities or bugs.

At PHPMD, our vocabulary is that we define rules and we raise violations of those rules, such violations have to be treated as code smells => to be checked in the context of your project.

That being said, we would agree to review and merge relevant merge-requests that would check if a parameter is required by an implemented interface.
